File: Data Structures & Algorithms/swim-in-rising-water/submission-9.java
class Solution {
    boolean[][] vis;
    int ROWS,COLS,res=0;
    PriorityQueue<int[]> minHeap = new PriorityQueue<>(
        (a,b)->Integer.compare(a[0],b[0])
    );
    int[][] directions = {{0,1},{1,0},{0,-1},{-1,0}};
    public int swimInWater(int[][] grid) {
        ROWS = grid.length;
        COLS = grid[0].length;
        vis = new boolean[ROWS][COLS];
        minHeap.offer(new int[] {0,0,0});
        vis[0][0] = true;
        while(!minHeap.isEmpty()){
            int[] swim = minHeap.poll();
            int row = swim[1], col = swim[2];
            res = Math.max(res,grid[row][col]);
            if(row == ROWS-1 && col == COLS-1)
                return res;
            for(int[] dir: directions) {
                int neiRow = row +dir[0], neiCol = col+dir[1];
                if(neiRow<0 || neiRow== ROWS || neiCol<0 || 
                    neiCol == COLS || vis[neiRow][neiCol]) {
                    continue;
                }
                minHeap.offer(new int[] {grid[neiRow][neiCol],neiRow,neiCol});
                vis[neiRow][neiCol] = true;
            }
        }
        return -1;
    }
}

File: Data Structures & Algorithms/islands-and-treasure/submission-9.py
class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        q = []
        M,N = len(grid),len(grid[0])
        vis = set()
        for i in range(M):
            for j in range(N):
                if grid[i][j] == 0:
                    q.append([i,j])
                    vis.add((i,j))
        dist = 0
        dir = [[0,1],[1,0],[0,-1],[-1,0]]
        while q:
            nextQ = []
            while q:
                i,j = q.pop()
                grid[i][j] = dist
                for dx,dy in dir:
                    x,y = i+dx,j+dy
                    if x<0 or y<0 or x==M or y ==N or grid[x][y] != 2**31 -1 or (x,y) in vis:
                        continue
                    vis.add((x,y))
                    nextQ.append([x,y])
            q = nextQ
            dist +=1
        
                    


                
        